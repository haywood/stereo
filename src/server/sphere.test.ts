import Sphere from "./sphere";
import { norm, equal } from "mathjs";

describe("Sphere", () => {
  describe("d=2", () => {
    it("generates a circle with 4 points", () => {
      const points = sphere(2, 1, 4);

      expect(
        equal(Array.from(points), [
          [-1, 0],
          [0, -1],
          [1, 0],
          [0, 1]
        ])
      ).toBeTruthy();
    });

    it("generates a circle with 10 points", () => {
      const points = sphere(2, 1, 10);

      expect(points).toMatchSnapshot();
    });
  });

  describe("d=3", () => {
    it("generates a sphere with 4 points", () => {
      const points = sphere(3, 1, 4);

      for (const p of points) {
        expect(norm(p)).toBeLessThanOrEqual(1);
      }

      expect(
        equal(Array.from(points), [
          [-1, 0, 0],
          [-1, 0, 0],
          [0, -1, 0],
          [1, 0, 0]
        ])
      ).toBeTruthy();
    });

    it("generates a sphere with 9 points", () => {
      const points = sphere(3, 1, 9);


      for (const p of points) {
        expect(norm(p)).toBeCloseTo(1, 10);
      }
      expect(points).toMatchSnapshot();
    });

    it("generates a sphere with 100 points", () => {
      const points = sphere(3, 1, 100);


      for (const p of points) {
        expect(norm(p)).toBeCloseTo(1, 10);
      }
      expect(points).toMatchSnapshot();
    });
  });

  function sphere(d, r, n) {
    return Array.from(new Sphere(d, r).sample(n));
  }
});
