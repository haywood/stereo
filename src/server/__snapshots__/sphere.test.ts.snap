// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sphere d=2 generates a circle with 10 points 1`] = `
Array [
  Array [
    -0.8090169943749473,
    0.5877852522924732,
  ],
  Array [
    -0.30901699437494734,
    0.9510565162951536,
  ],
  Array [
    0.30901699437494745,
    0.9510565162951535,
  ],
  Array [
    0.8090169943749475,
    0.5877852522924731,
  ],
  Array [
    1,
    0,
  ],
  Array [
    0.8090169943749475,
    -0.5877852522924731,
  ],
  Array [
    0.30901699437494745,
    -0.9510565162951535,
  ],
  Array [
    -0.30901699437494734,
    -0.9510565162951536,
  ],
  Array [
    -0.8090169943749473,
    -0.5877852522924732,
  ],
  Array [
    -1,
    -1.2246467991473532e-16,
  ],
]
`;

exports[`Sphere d=3 generates a sphere with 9 points 1`] = `
Array [
  Array [
    0.25000000000000033,
    0.4330127018922195,
    0.8660254037844385,
  ],
  Array [
    0.2500000000000001,
    0.43301270189221913,
    -0.8660254037844387,
  ],
  Array [
    -0.5000000000000003,
    -0.8660254037844385,
    -1.2246467991473532e-16,
  ],
  Array [
    0.2500000000000001,
    -0.43301270189221963,
    0.8660254037844385,
  ],
  Array [
    0.2499999999999999,
    -0.43301270189221924,
    -0.8660254037844387,
  ],
  Array [
    -0.4999999999999999,
    0.8660254037844387,
    -1.2246467991473532e-16,
  ],
  Array [
    -0.5000000000000003,
    -6.12323399573677e-17,
    0.8660254037844385,
  ],
  Array [
    -0.4999999999999999,
    -6.123233995736765e-17,
    -0.8660254037844387,
  ],
  Array [
    1,
    1.2246467991473532e-16,
    -1.2246467991473532e-16,
  ],
]
`;

exports[`Sphere d=3 generates a sphere with 100 points 1`] = `
Array [
  Array [
    0.6545084971874736,
    -0.4755282581475768,
    0.5877852522924732,
  ],
  Array [
    0.24999999999999992,
    -0.1816356320013402,
    0.9510565162951536,
  ],
  Array [
    -0.25,
    0.18163563200134028,
    0.9510565162951535,
  ],
  Array [
    -0.6545084971874736,
    0.4755282581475769,
    0.5877852522924731,
  ],
  Array [
    -0.8090169943749473,
    0.5877852522924732,
    0,
  ],
  Array [
    -0.6545084971874736,
    0.4755282581475769,
    -0.5877852522924731,
  ],
  Array [
    -0.25,
    0.18163563200134028,
    -0.9510565162951535,
  ],
  Array [
    0.24999999999999992,
    -0.1816356320013402,
    -0.9510565162951536,
  ],
  Array [
    0.6545084971874736,
    -0.4755282581475768,
    -0.5877852522924732,
  ],
  Array [
    0.8090169943749473,
    -0.5877852522924732,
    -1.2246467991473532e-16,
  ],
  Array [
    0.24999999999999992,
    -0.7694208842938133,
    0.5877852522924732,
  ],
  Array [
    0.09549150281252623,
    -0.2938926261462365,
    0.9510565162951536,
  ],
  Array [
    -0.09549150281252627,
    0.2938926261462366,
    0.9510565162951535,
  ],
  Array [
    -0.24999999999999994,
    0.7694208842938134,
    0.5877852522924731,
  ],
  Array [
    -0.30901699437494734,
    0.9510565162951536,
    0,
  ],
  Array [
    -0.24999999999999994,
    0.7694208842938134,
    -0.5877852522924731,
  ],
  Array [
    -0.09549150281252627,
    0.2938926261462366,
    -0.9510565162951535,
  ],
  Array [
    0.09549150281252623,
    -0.2938926261462365,
    -0.9510565162951536,
  ],
  Array [
    0.24999999999999992,
    -0.7694208842938133,
    -0.5877852522924732,
  ],
  Array [
    0.30901699437494734,
    -0.9510565162951536,
    -1.2246467991473532e-16,
  ],
  Array [
    -0.25,
    -0.7694208842938133,
    0.5877852522924732,
  ],
  Array [
    -0.09549150281252627,
    -0.29389262614623646,
    0.9510565162951536,
  ],
  Array [
    0.0954915028125263,
    0.29389262614623657,
    0.9510565162951535,
  ],
  Array [
    0.25000000000000006,
    0.7694208842938134,
    0.5877852522924731,
  ],
  Array [
    0.30901699437494745,
    0.9510565162951535,
    0,
  ],
  Array [
    0.25000000000000006,
    0.7694208842938134,
    -0.5877852522924731,
  ],
  Array [
    0.0954915028125263,
    0.29389262614623657,
    -0.9510565162951535,
  ],
  Array [
    -0.09549150281252627,
    -0.29389262614623646,
    -0.9510565162951536,
  ],
  Array [
    -0.25,
    -0.7694208842938133,
    -0.5877852522924732,
  ],
  Array [
    -0.30901699437494745,
    -0.9510565162951535,
    -1.2246467991473532e-16,
  ],
  Array [
    -0.6545084971874736,
    -0.47552825814757677,
    0.5877852522924732,
  ],
  Array [
    -0.24999999999999994,
    -0.18163563200134017,
    0.9510565162951536,
  ],
  Array [
    0.25000000000000006,
    0.18163563200134025,
    0.9510565162951535,
  ],
  Array [
    0.6545084971874737,
    0.4755282581475768,
    0.5877852522924731,
  ],
  Array [
    0.8090169943749475,
    0.5877852522924731,
    0,
  ],
  Array [
    0.6545084971874737,
    0.4755282581475768,
    -0.5877852522924731,
  ],
  Array [
    0.25000000000000006,
    0.18163563200134025,
    -0.9510565162951535,
  ],
  Array [
    -0.24999999999999994,
    -0.18163563200134017,
    -0.9510565162951536,
  ],
  Array [
    -0.6545084971874736,
    -0.47552825814757677,
    -0.5877852522924732,
  ],
  Array [
    -0.8090169943749475,
    -0.5877852522924731,
    -1.2246467991473532e-16,
  ],
  Array [
    -0.8090169943749473,
    0,
    0.5877852522924732,
  ],
  Array [
    -0.30901699437494734,
    0,
    0.9510565162951536,
  ],
  Array [
    0.30901699437494745,
    0,
    0.9510565162951535,
  ],
  Array [
    0.8090169943749475,
    0,
    0.5877852522924731,
  ],
  Array [
    1,
    0,
    0,
  ],
  Array [
    0.8090169943749475,
    0,
    -0.5877852522924731,
  ],
  Array [
    0.30901699437494745,
    0,
    -0.9510565162951535,
  ],
  Array [
    -0.30901699437494734,
    0,
    -0.9510565162951536,
  ],
  Array [
    -0.8090169943749473,
    0,
    -0.5877852522924732,
  ],
  Array [
    -1,
    0,
    -1.2246467991473532e-16,
  ],
  Array [
    -0.6545084971874736,
    0.47552825814757677,
    0.5877852522924732,
  ],
  Array [
    -0.24999999999999994,
    0.18163563200134017,
    0.9510565162951536,
  ],
  Array [
    0.25000000000000006,
    -0.18163563200134025,
    0.9510565162951535,
  ],
  Array [
    0.6545084971874737,
    -0.4755282581475768,
    0.5877852522924731,
  ],
  Array [
    0.8090169943749475,
    -0.5877852522924731,
    0,
  ],
  Array [
    0.6545084971874737,
    -0.4755282581475768,
    -0.5877852522924731,
  ],
  Array [
    0.25000000000000006,
    -0.18163563200134025,
    -0.9510565162951535,
  ],
  Array [
    -0.24999999999999994,
    0.18163563200134017,
    -0.9510565162951536,
  ],
  Array [
    -0.6545084971874736,
    0.47552825814757677,
    -0.5877852522924732,
  ],
  Array [
    -0.8090169943749475,
    0.5877852522924731,
    -1.2246467991473532e-16,
  ],
  Array [
    -0.25,
    0.7694208842938133,
    0.5877852522924732,
  ],
  Array [
    -0.09549150281252627,
    0.29389262614623646,
    0.9510565162951536,
  ],
  Array [
    0.0954915028125263,
    -0.29389262614623657,
    0.9510565162951535,
  ],
  Array [
    0.25000000000000006,
    -0.7694208842938134,
    0.5877852522924731,
  ],
  Array [
    0.30901699437494745,
    -0.9510565162951535,
    0,
  ],
  Array [
    0.25000000000000006,
    -0.7694208842938134,
    -0.5877852522924731,
  ],
  Array [
    0.0954915028125263,
    -0.29389262614623657,
    -0.9510565162951535,
  ],
  Array [
    -0.09549150281252627,
    0.29389262614623646,
    -0.9510565162951536,
  ],
  Array [
    -0.25,
    0.7694208842938133,
    -0.5877852522924732,
  ],
  Array [
    -0.30901699437494745,
    0.9510565162951535,
    -1.2246467991473532e-16,
  ],
  Array [
    0.24999999999999992,
    0.7694208842938133,
    0.5877852522924732,
  ],
  Array [
    0.09549150281252623,
    0.2938926261462365,
    0.9510565162951536,
  ],
  Array [
    -0.09549150281252627,
    -0.2938926261462366,
    0.9510565162951535,
  ],
  Array [
    -0.24999999999999994,
    -0.7694208842938134,
    0.5877852522924731,
  ],
  Array [
    -0.30901699437494734,
    -0.9510565162951536,
    0,
  ],
  Array [
    -0.24999999999999994,
    -0.7694208842938134,
    -0.5877852522924731,
  ],
  Array [
    -0.09549150281252627,
    -0.2938926261462366,
    -0.9510565162951535,
  ],
  Array [
    0.09549150281252623,
    0.2938926261462365,
    -0.9510565162951536,
  ],
  Array [
    0.24999999999999992,
    0.7694208842938133,
    -0.5877852522924732,
  ],
  Array [
    0.30901699437494734,
    0.9510565162951536,
    -1.2246467991473532e-16,
  ],
  Array [
    0.6545084971874736,
    0.4755282581475768,
    0.5877852522924732,
  ],
  Array [
    0.24999999999999992,
    0.1816356320013402,
    0.9510565162951536,
  ],
  Array [
    -0.25,
    -0.18163563200134028,
    0.9510565162951535,
  ],
  Array [
    -0.6545084971874736,
    -0.4755282581475769,
    0.5877852522924731,
  ],
  Array [
    -0.8090169943749473,
    -0.5877852522924732,
    0,
  ],
  Array [
    -0.6545084971874736,
    -0.4755282581475769,
    -0.5877852522924731,
  ],
  Array [
    -0.25,
    -0.18163563200134028,
    -0.9510565162951535,
  ],
  Array [
    0.24999999999999992,
    0.1816356320013402,
    -0.9510565162951536,
  ],
  Array [
    0.6545084971874736,
    0.4755282581475768,
    -0.5877852522924732,
  ],
  Array [
    0.8090169943749473,
    0.5877852522924732,
    -1.2246467991473532e-16,
  ],
  Array [
    0.8090169943749473,
    9.907600726170915e-17,
    0.5877852522924732,
  ],
  Array [
    0.30901699437494734,
    3.7843667304341493e-17,
    0.9510565162951536,
  ],
  Array [
    -0.30901699437494745,
    -3.7843667304341506e-17,
    0.9510565162951535,
  ],
  Array [
    -0.8090169943749475,
    -9.907600726170916e-17,
    0.5877852522924731,
  ],
  Array [
    -1,
    -1.2246467991473532e-16,
    0,
  ],
  Array [
    -0.8090169943749475,
    -9.907600726170916e-17,
    -0.5877852522924731,
  ],
  Array [
    -0.30901699437494745,
    -3.7843667304341506e-17,
    -0.9510565162951535,
  ],
  Array [
    0.30901699437494734,
    3.7843667304341493e-17,
    -0.9510565162951536,
  ],
  Array [
    0.8090169943749473,
    9.907600726170915e-17,
    -0.5877852522924732,
  ],
  Array [
    1,
    1.2246467991473532e-16,
    -1.2246467991473532e-16,
  ],
]
`;
